bugbug/similarity.py

```diff
@@ -268,134 +267,95 @@ def __init__(self, cut_off=0.2, cleanup_urls=True, nltk_tokenizer=False):
-            d = np.zeros(vocab_len, dtype=np.double)
-            nbow = dictionary.doc2bow(document)
-            doc_len = len(document)
-            for idx, freq in nbow:
-                d[idx] = freq / float(doc_len)
-            return d
-
-        d1 = nbow(document1)
-        d2 = nbow(document2)
-
-        return emd(d1, d2, distance_matrix)
-
-    def calculate_all_distances(self, words):
-        return np.array(
-            1.0
-            - np.dot(
-                self.w2vmodel.wv.vectors_norm,
-                self.w2vmodel.wv.vectors_norm[
-                    [self.w2vmodel.wv.vocab[word].index for word in words]
-                ].transpose(),
-            ),
-            dtype=np.double,
-        )
+        self.dictionary = Dictionary(self.corpus)
+        self.tfidf = TfidfModel(dictionary=self.dictionary)
 
     def get_similar_bugs(self, query):
 
-        words = self.text_preprocess(self.get_text(query))
-        words = [word for word in words if word in self.w2vmodel.wv.vocab]
-
-        all_distances = self.calculate_all_distances(words)
-
-        distances = []
-        for i in range(len(self.corpus)):
-            cleaned_corpus = [
-                word for word in self.corpus[i] if word in self.w2vmodel.wv.vocab
-            ]
-            indexes = [self.w2vmodel.wv.vocab[word].index for word in cleaned_corpus]
-            if len(indexes) != 0:
-                word_dists = all_distances[indexes]
-                rwmd = max(
-                    np.sum(np.min(word_dists, axis=0)),
-                    np.sum(np.min(word_dists, axis=1)),
+        query = self.text_preprocess(self.get_text(query))
+        words = [
+            word for word in set(chain(query, *self.corpus)) if word in self.w2vmodel.wv
+        ]
+        indices, words = zip(
+            *sorted(
+                (
+                    (index, word)
+                    for (index, _), word in zip(self.dictionary.doc2bow(words), words)
                 )
-
-                distances.append((self.bug_ids[i], rwmd))
-
-        distances.sort(key=lambda v: v[1])
-
-        confirmed_distances_ids = []
-        confirmed_distances = []
-
-        for i, (doc_id, rwmd_distance) in enumerate(distances):
-
-            if (
-                len(confirmed_distances) >= 10
-                and rwmd_distance > confirmed_distances[10 - 1]
-            ):
-                break
-
-            doc_words_clean = [
-                word
-                for word in self.corpus[self.bug_ids.index(doc_id)]
-                if word in self.w2vmodel.wv.vocab
+            )
+        )
+        query = dict(self.tfidf[self.dictionary.doc2bow(query)])
+        query = [
+            (new_index, query[dict_index])
+            for new_index, dict_index in enumerate(indices)
+            if dict_index in query
+        ]
+        documents = [
+            dict(self.tfidf[self.dictionary.doc2bow(document)])
+            for document in self.corpus
+        ]
```

Comment:
I wonder if this couldn't be done just once in the constructor and cached, it doesn't seem to use any data from the query itself.
How long does `get_similar_bugs` take if you move this out? Is it a measurable improvement?

Reply:
With the current implementation, it takes 5.4607 seconds to test on the first 10 bugs.
Taking it ti the constructor (making this an attribute) actually increases the time to 5.7524 seconds.